java/arcs/core/storage/handle/RawEntityDereferencer.kt

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.storage.handle
+
+import arcs.core.crdt.CrdtEntity
+import arcs.core.data.EntityType
+import arcs.core.data.RawEntity
+import arcs.core.data.Schema
+import arcs.core.storage.Dereferencer
+import arcs.core.storage.EntityActivationFactory
+import arcs.core.storage.ProxyCallback
+import arcs.core.storage.ProxyMessage
+import arcs.core.storage.Reference
+import arcs.core.storage.StorageMode
+import arcs.core.storage.Store
+import arcs.core.storage.StoreOptions
+import arcs.core.util.TaggedLog
+import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+/**
+ * [Dereferencer] to use when de-referencing a [Reference] to an [Entity].
+ *
+ * [Handle] implementations should inject this into any [Reference] objects they encounter.
+ */
+class RawEntityDereferencer(
+    private val schema: Schema,
+    private val entityActivationFactory: EntityActivationFactory? = null
+) : Dereferencer<RawEntity> {
+    private val log = TaggedLog { "Dereferencer(${schema.names})" }
+
+    override suspend fun dereference(
```

Comment:
I'm kind of worried about the performance characteristics of this model if someone's processing say a join of two entities.  In the Java ORM world, this reminds me of the 'N+1' selects problem.  In the ORM world, you had configuration options to use a real join and populate the related entities without additional queries (the data needed for dereferencing would be pre-fetched in the query)

Granted, for the examples we're looking at, the size of our data is small, but for large collections of relations, it could be problematic?

Reply:
Good point. It absolutely would be problematic in that situation, this implementation assumes the user is okay with accepting that overhead.

 in the future I think we will be able to optimize how we go about dereferencing references (even: returning fully-dereferenced entities/collections/singletons on the first query).